internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,301 @@
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		logger.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
+	sl.client = client
+
+	if sl.mList == nil {
+		return nil
+	}
+
+	go sl.memberlistWatchEvents()
+	go sl.mlJoinLoop()
+
+	return nil
+}
+
+// updateMlIPs runs forever updating the sl.mlSpeakersIPs slice with the
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
+	for {
+		// Create the ticker at the start of each loop, so we wait at
+		// least the ticker interval before the ticker case is true.
+		// If we create it outside of the loop, it will tick every 5
+		// minutes and if the blocking call happens to take longer, the
+		// next tick will be really soon and we may overload the API
+		// server.
+		ticker := time.NewTicker(5 * time.Minute)
+
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			// This blocks until the apiserver responds.
+			iplist, err := sl.mlSpeakers()
+			if err != nil {
+				sl.l.Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
+				continue
+			}
+
+			sl.mlMux.Lock()
+			sl.mlSpeakersIPs = iplist
+			sl.mlMux.Unlock()
+		}
+	}
+}
+
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response of the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context or timeout.
+	// TODO: When updating client-go, this code can be simplified by using a
+	// context in the following way:
+	// If we use a context, we can specify a timeout and there is no need to
+	// have the speakersIPsLoop() running in the background. This can just
+	// be called from mlJoin() with a context (timeout) and if doesn't
+	// answer, use sl.mlSpeakersIPs.  If it doesn't time out, just update
+	// sl.mlSpeakersIPs with the answer and continue to do the join.
+	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
+	if err != nil {
+		return nil, err
+	}
+
+	return iplist, nil
+}
+
+func (sl *SpeakerList) mlJoinLoop() {
+	iplist, err := sl.mlSpeakers()
+	if err != nil {
+		sl.l.Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
+		iplist = nil
+	}
+
+	// Initialize sl.mlSpeakersIPs.
+	sl.mlMux.Lock()
+	sl.mlSpeakersIPs = iplist
+	sl.mlMux.Unlock()
+
+	// Update mlSpeakersIPs in the background.
+	go sl.speakersIPsLoop()
+
+	// Every one minute, try to rejoin.
+	// This will join nodes that leave the cluster only for a few seconds every 1
+	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
+	// take a while longer.
+	ticker := time.NewTicker(1 * time.Minute)
```

Comment:
Oh, maybe that is simpler. Good point. Let's see if we want to watch for k8s events or not, and if we are going this way I'll check if that makes the code simpler. Thanks! :)

Reply:
If we watch for k8s objects it's going to be closer and closer to my PR or am I missing something ?

Reply:
I think rata@ doesn't want to introduce extra resource objects in manifests (i.e. the extra service in your PR), which is hard to remove if we ever want to remove it. That's the key difference I see it.

Reply:
It's not a problem if we don't remove it, and I'm providing a clean way to remove it ...
Right now we only have pod list permission, so we can't change that for a patch release.

Reply:
Yes, I don't want to include a service if we will remove it in the next release. Watch permissions, if and only if it is more clear that we will keep them, seems better, IMHO.

Reply:
Still, can't change the manifests in a patch release IMO.

I still believe that we should use the best tool for the job (here watching service), but if that's the only blocking point, I can have the k8s client go watching pods in SpeakerList

Reply:
Switched to `time.After()`, it indeed makes it way simpler. Thanks!

Reply:
Advantage of ticker is that it run every X exactly, so it sometimes make things easier to investigate weird behavior (I'm on my phone so can't see code before / after)

Reply:
Not really. If the ticker is every X and when it ticks you call a function that takes 2X time, then you will call the function the first time and receive the second time after a few milliseconds.

The idea here is to avoid doing that, just make sure to not call more than every 5 minutes. 

If the API server is down, for example, your calls can take long and you might end up calling the API server more than expected. I think using `time.After()` is quite simpler here.